### PR TITLE
__getitem__ and __setitem__ support for structured arrays and scalars

### PR DESCRIPTION
**Pull request by pitrou**
This adds a dedicated IR rewrite pass before type inference, attempting to detect getitem and setitem lookups using compile-time constants and replacing them with dedicated IR expressions/statements.

### CODE REVIEW

**Comment by seibert** [comment]
@sklam: Can you review this?

File: numba/tests/test_record_dtype.py [hunk 1]
```diff
@@ -28,6 +38,15 @@ def get_b_subarray(ary, i):
 def get_c_subarray(ary, i):
     return ary.c[i]
 
+def make_getitem_subarray(item):
+    # This also exercises constant lookup from a closure variable
+    def get_xx_subarray(ary, i):
+        return ary[i][item]
```
Comment (sklam) [inline numba/tests/test_record_dtype.py]: I think this should be `ary[item][i]`

Comment (pitrou) [inline numba/tests/test_record_dtype.py]: Nice catch. Yes, you are right, this isn't testing what it should test.

**Comment by seibert** [comment]
Looks good.  Thanks!

**Pull request merged**